src/solvers/dgmulti/shock_capturing.jl

```diff
@@ -0,0 +1,284 @@
+  FToF = mesh.md.FToF # num_faces x num_elements matrix
+  EToE = similar(FToF)
+  for e in axes(FToF, 2)
+    for f in axes(FToF, 1)
+      neighbor_face_index = FToF[f, e]
+
+      # reverse-engineer element index from face. Assumes all elements
+      # have the same number of faces.
+      neighbor_element_index = ((neighbor_face_index - 1) ÷ dg.basis.num_faces) + 1
+      EToE[f, e] = neighbor_element_index
+    end
+  end
+
+  # create sparse hybridized operators for low order scheme
+  Qrst, E = StartUpDG.sparse_low_order_SBP_operators(dg.basis)
+  Brst = map(n -> Diagonal(n .* dg.basis.wf), dg.basis.nrstJ)
+  sparse_hybridized_SBP_operators = map((Q, B) -> 0.5 * [Q-Q' E'*B; -B*E zeros(size(B))], Qrst, Brst)
+
+  # Find the joint sparsity pattern of the entire matrix. We store the sparsity pattern as
+  # a transpose for faster iteration through the rows.
+  sparsity_pattern = transpose(sum(map(A -> abs.(A), sparse_hybridized_SBP_operators)) .> 100 * eps())
+
+  return (; element_ids_dg, element_ids_dgfv, sparse_hybridized_SBP_operators, sparsity_pattern, EToE)
+end
+
+
+# this method is used when the indicator is constructed as for shock-capturing volume integrals
+function create_cache(::Type{IndicatorHennemannGassner}, equations::AbstractEquations,
+                      basis::RefElemData{NDIMS}) where NDIMS
+
+  alpha = Vector{real(basis)}()
+  alpha_tmp = similar(alpha)
+
+  A = Vector{real(basis)}
+  indicator_threaded  = [A(undef, nnodes(basis)) for _ in 1:Threads.nthreads()]
+  modal_threaded      = [A(undef, nnodes(basis)) for _ in 1:Threads.nthreads()]
+
+  # initialize inverse Vandermonde matrices at Gauss-Legendre nodes
+  (; N) = basis
+  lobatto_node_coordinates_1D, _ = StartUpDG.gauss_lobatto_quad(0, 0, N)
+  VDM_1D = StartUpDG.vandermonde(Line(), N, lobatto_node_coordinates_1D)
+  inverse_vandermonde = SimpleKronecker(NDIMS, inv(VDM_1D))
+
+  return (; alpha, alpha_tmp, indicator_threaded, modal_threaded, inverse_vandermonde)
+end
+
+
+function (indicator_hg::IndicatorHennemannGassner)(u, mesh::DGMultiMesh,
+                                                   equations, dg::DGMulti{NDIMS}, cache;
+                                                   kwargs...) where {NDIMS}
+  (; alpha_max, alpha_min, alpha_smooth, variable) = indicator_hg
+  (; alpha, alpha_tmp, indicator_threaded, modal_threaded, inverse_vandermonde) = indicator_hg.cache
+
+  resize!(alpha, nelements(mesh, dg))
+  if alpha_smooth
+    resize!(alpha_tmp, nelements(mesh, dg))
+  end
+
+  # magic parameters
+  threshold = 0.5 * 10^(-1.8 * (dg.basis.N + 1)^0.25)
+  parameter_s = log((1 - 0.0001) / 0.0001)
+
+  @threaded for element in eachelement(mesh, dg)
+    indicator = indicator_threaded[Threads.threadid()]
+    modal_ = modal_threaded[Threads.threadid()]
+
+    # Calculate indicator variable at interpolation (Lobatto) nodes.
+    # TODO: calculate indicator variables at Gauss nodes or using `cache.entropy_projected_u_values`
```

Comment:
Open TODO?

Reply:
Yup. I think it makes more sense to address this in a future PR, the refactoring was uglier than I expected.

Reply:
Ok. Do you track it in the corresponding issue?

Reply:
Good point, adding now.